### ISSUE TITLE
Create a class hierarchy for particles

### ISSUE THREAD

**Issue by namurphy** [PlasmaPy/PlasmaPy#687]
Right now, ions, electrons, leptons, baryons, and so on are all represented by instances of the `plasmapy.atomic.Particle` class.  This class contains a method, `Particle.is_category`, and can be used to tell whether or not a `Particle` instance is a particular category, like:
```Python
>>> electron = Particle('e-')
>>> electron.is_category('lepton')
True
>>> electron.is_category('baryon')
```
My thinking is that it would be more Pythonic to be able to do something like:
```Python
>>> isinstance(electron, Lepton)
True
>>> isinstance(proton, Baryon)
False
```
We could keep `Particle` as the factory for particles, and define `GenericParticle` as the abstract base class for particles.  Then we could have `Lepton`, `Baryon`, `Ion`, `NeutralAtom`, `AntiParticle` and whatever else inherit from `GenericParticle` and each other, as necessary.  `CustomParticle` as described in #447 would inherit from `GenericParticle` too.

A potential downside is that we would probably end up using multiple inheritance, which can be messy to modify.  With that said, the particle hierarchy exists in nature, so it would be sensible to reflect this in the code.  This idea warrants further discussion before we possibly implement it, so we should think of this as a proposal for the time being.

**Comment by namurphy** [comment]
Oh...another benefit from doing this is that we would be able to do things like this:

```Python
@particle_input
def ion_inertial_length(ion: Ion, ...):
    return d_i
```

We could then have `@particle_input` enforce that `ion` is in fact an `Ion`.  Right now I think we do that a little bit based on what the argument's name is (i.e., `@particle_input` will check that arguments named `ion` are actually ions, unless I'm misremembering).  

We probably would also want `ChargedParticle`, `NeutralParticle`, and `NeutralAtom` classes in the hierarchy too.

**Comment by goodab** [comment]
I agree that the current implementation of `Particle` is pretty dense and hard to extend. 

A pretty simple class hierarchy with `Particle` as the base class that `SubatomicParticle` and `ElementParticle` both subclass and an additional subclass `Isotope` to `ElementParticle`, handles most of the shared functionality without much code duplication. It runs into an issue with defining the proton, since it is both a `SubatomicParticle` and an `Isotope`, but that can be fixed with a specific `Proton` subclass with multiple inheritance.

An alternative to the class structure is to use composition for the particle properties. This makes the proton edge case more attractive at the cost of making everything else clunkier.

As for type-checking, `@particle_input` does a pretty good job of using the type annotation tools to enforce rules on the inputs and switching over to a bunch of mixin-classes to enforce those rules doesn't seem to add much besides being able to use isinstance().

**Comment by goodab** [comment]
Actually, we might be able to swing something with `ABCMeta` that allows you to use `isinstance` to replace `is_category` while also making that behavior clearer. [This SO thread](https://stackoverflow.com/questions/32124683/duck-typing-isinstance-subclasshook-overriding-and-types-methodtype?lq=1) demonstrates using `__subclasshook__` in `ABCMeta` to dynamically determine type.

**Comment by goodab** [comment]
Here's a solution that provides conditional type checking via `isinstance` without having to deal with multiple inheritance.

```
import abc
from abc import ABCMeta

class ConditionalType(ABCMeta):
  def __instancecheck__(cls, other):
    return cls.condition(other)
  
  @classmethod
  @abc.abstractmethod
  def condition(cls, other) -> bool:
    pass
    
class ChargedParticle(metaclass=ConditionalType):
  @classmethod
  def condition(cls, other) -> bool:
    try:
      return other.charge != 0 and isinstance(other, Particle)
    except AttributeError:
      return False

class Particle():
  def __init__(self, charge):
    self.charge = charge

assert isinstance(Particle(1), ChargedParticle)
assert not isinstance(Particle(0), ChargedParticle)
```

As shown in the example, this overrides isinstance behavior as expected, so we could modify @particle_input to handle the type check as well. Unfortunately, I don't think it will work with static type checkers.

**Comment by namurphy** [comment]
@goodab - Thank you!  That does look like a very valid way of doing this.  I had been wondering if there was a way we could do something like this with metaclasses, but I haven't had enough experience with them to know.

> Unfortunately, I don't think it will work with static type checkers.

At the moment, we're unable to use mypy anyway because we use annotations from `astropy.units`.   (#268).  I'm also wondering if there is a workaround that we could implement using mypy (since ultimately it would be really helpful to have a static type check for each pull request).

**Comment by namurphy** [comment]
We discussed this in today's community meeting, and the first step will be to create an `AbstractParticle` class, and then subclass it for `Particle` and `CustomParticle`.  We also wanted to create a dimensionless representation of a particle.  We didn't decide on a name, but it should be something like `DimensionlessParticle`, `NormalizedParticle`, `SimulationParticle`, and `ScaledParticle` :snake:.

**Comment by namurphy** [comment]
An update...#755 created `AbstractParticle`, `DimensionlessParticle`, and `CustomParticle`.  

I'm still thinking that this would be a really helpful feature for `plasmapy.particles`.  In particular, there are a lot of attributes to `Particle` that are *only* relevant to elements (`standard_atomic_weight`), ions (`roman_symbol`), fundamental particles (`spin`), etc.  Having a particle hierarchy could reduce clutter in the namespace.
